flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/container/FlinkImageBuilder.java

```diff
@@ -0,0 +1,293 @@
+        this.startupCommand = "flink/bin/jobmanager.sh start-foreground && tail -f /dev/null";
+        this.imageNameSuffix = "jobmanager";
+        return this;
+    }
+
+    /** Use this image for building a TaskManager. */
+    public FlinkImageBuilder asTaskManager() {
+        checkStartupCommandNotSet();
+        this.startupCommand = "flink/bin/taskmanager.sh start-foreground && tail -f /dev/null";
+        this.imageNameSuffix = "taskmanager";
+        return this;
+    }
+
+    /** Use a custom command for starting up the container. */
+    public FlinkImageBuilder useCustomStartupCommand(String command) {
+        checkStartupCommandNotSet();
+        this.startupCommand = command;
+        this.imageNameSuffix = "custom";
+        return this;
+    }
+
+    /** Build the image. */
+    public ImageFromDockerfile build() throws ImageBuildException {
+        sanityCheck();
+        final String finalImageName = imageName + "-" + imageNameSuffix;
+        try {
+            // Build base image first
+            buildBaseImage(flinkDist);
+            final Path tempDirectory =
+                    Files.createTempDirectory("flink-dist-build-" + UUID.randomUUID());
+            final Path flinkConfFile = createTemporaryFlinkConfFile(tempDirectory);
+            final Path log4jPropertiesFile = createTemporaryLog4jPropertiesFile(tempDirectory);
+            // Copy flink-conf.yaml into image
+            filesToCopy.put(
+                    flinkConfFile,
+                    Paths.get("flink", "conf", GlobalConfiguration.FLINK_CONF_FILENAME));
+            filesToCopy.put(
+                    log4jPropertiesFile, Paths.get("flink", "conf", LOG4J_PROPERTIES_FILENAME));
+
+            final ImageFromDockerfile image =
+                    new ImageFromDockerfile(finalImageName)
+                            .withDockerfileFromBuilder(
+                                    builder -> {
+                                        // Build from base image
+                                        builder.from(FLINK_BASE_IMAGE_NAME);
+                                        // Copy files into image
+                                        filesToCopy.forEach(
+                                                (from, to) ->
+                                                        builder.copy(to.toString(), to.toString()));
+                                        builder.cmd(startupCommand);
+                                    });
+            filesToCopy.forEach((from, to) -> image.withFileFromPath(to.toString(), from));
+            return image;
+        } catch (Exception e) {
+            throw new ImageBuildException(
+                    String.format("Failed to build Flink image \"%s\"", finalImageName), e);
+        }
+    }
+
+    // ----------------------- Helper functions -----------------------
+
+    private void buildBaseImage(Path flinkDist) throws TimeoutException {
+        if (!baseImageExists()) {
```

Comment:
Nit: I am a fan of handling of all conditional cases explicitly. I think it makes reading the code easier and removes one indentation.

```java
if (baseImageExists()) {
    return;
}
...
```